### PR TITLE
Create new `circomlibjs` package

### PR DESCRIPTION
**Pull request by cedoor**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/circomlibjs/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+    multiDel(dels: any[]): void;
+
+    multiGet(keys: any[]): any;
+
+    multiIns(inserts: any[]): void;
+
+    setRoot(root: any): void;
+}
+
+export interface BabyJub {
```
Comment (stanbar) [inline types/circomlibjs/index.d.ts:49]: I implemented typed wrapper around BabyJub for my own usage. I figured out the following types:

```typescript
type BabyJubPoint = [Uint8Array, Uint8Array]

const Generator: BabyJubPoint = babyjub.Generator
const Base8: BabyJubPoint = babyjub.Base8

function mulPointEscalar(p: BabyJubPoint, scalar: bigint|number): BabyJubPoint {
    return babyjub.mulPointEscalar(p, scalar)
}

function inCurve(p: BabyJubPoint): boolean {
    return babyjub.inCurve(p)
}

function inSubgroup(p: BabyJubPoint): boolean {
    return babyjub.inSubgroup(p)
}

function addPoint(p1: BabyJubPoint, p2: BabyJubPoint): BabyJubPoint {
    return babyjub.addPoint(p1, p2)
}
```

File: types/circomlibjs/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,343 @@
+// Type definitions for circomlibjs 0.1
+// Project: https://github.com/iden3/circomlibjs
+// Definitions by: cedoor <https://github.com/cedoor>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+export type Poseidon = (arr: any[], state?: any, nOut?: number) => any;
+
+export type Point = [any, any];
```
Comment (stanbar) [inline types/circomlibjs/index.d.ts:8]: I think you can constrain `export type Point = [any, any];`  with `export type Point = [Uint8Array, Uint8Array];`

File: types/circomlibjs/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+
+    multiGet(keys: any[]): any;
+
+    multiIns(inserts: any[]): void;
+
+    setRoot(root: any): void;
+}
+
+export interface BabyJub {
+    addPoint(a: any, b: any): Point;
```
Comment (stanbar) [inline types/circomlibjs/index.d.ts:50]:
```typescript
addPoint(p1: BabyJubPoint, p2: BabyJubPoint): BabyJubPoint
```

File: types/circomlibjs/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+
+    mulPointEscalar(base: any, e: any): Point;
```
Comment (stanbar) [inline types/circomlibjs/index.d.ts:52]:
```typescript
mulPointEscalar(p: BabyJubPoint, scalar: bigint|number): BabyJubPoint
```

File: types/circomlibjs/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+
+    inSubgroup(point: Point): boolean;
```
Comment (stanbar) [inline types/circomlibjs/index.d.ts:54]:
```typescript
inSubgroup(p: BabyJubPoint): boolean
```

File: types/circomlibjs/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+
+    inCurve(point: Point): boolean;
```
Comment (stanbar) [inline types/circomlibjs/index.d.ts:56]:
```typescript
inCurve(p: BabyJubPoint): boolean
```

**Review by stanbar — COMMENTED** [review]
I think you may constrain BabyJub parameters with `Uint8Array`. For me, it saved hours of debugging.

**Review by gabritto — CHANGES_REQUESTED** [review]
So, @stanbar left some comments that you might want to address, and type-wise I think everything looks ok.
Other than that, the original package is a dual package: it ships both ESM and CommonJS files/entry-points. The types here should reflect that.
In order to achieve this, we need a few things:
1. The DT package should have a `package.json` file, with `"private": true`.
1.1: The `package.json` needs to have `"type": "module"` to reflect the original `package.json`. This is needed to tell TS that .ts/.js files should be interpreted as ESM files.
1.2: The `package.json` also needs to have an exports map that matches the [original exports map](https://unpkg.com/browse/circomlibjs@0.1.7/package.json):

Original:
```json
{
	"exports": {
		"import": "./main.js",
		"require": "./build/main.cjs"
	}
}
```

On the DT's `package.json`:

```json
{
	"exports": {
		".": {
			"types": {
				"import": "./index.d.ts",
				"require": "./index.d.cts"
			}
		}
	}
}
```
2. Like the exports map suggests above, there should be two declaration files, one that is ESM (`index.d.ts`), and a CJS one (`index.d.cts`).
To avoid duplicating the types, since they seem to be the same for both of the ESM and CJS entry points of the original package, you can define the types in the `index.d.cts` file, and import and re-export everything in the `index.ts` file.

**Comment by cedoor** [comment]
> I think you may constrain BabyJub parameters with `Uint8Array`. For me, it saved hours of debugging.

Hey @stanbar, I've used `any` for them because the point values and some of the other parameters can actually be `number`, `string` (string number), `bigint` or `Uint8Array` most of the times. 

`F.mul` (within `babyJub.addPoint`) can also support those types, as you can see here: https://github.com/iden3/ffjavascript/blob/559d532f5482f01191385e8f485ba35183abf1c4/src/f1field.js#L89.

One solution could be to create a type like (paying attention to those parameters that can only be Uint8Array):

```ts
export type BigNumberish = string | bigint | number | Uint8Array
```

**Comment by cedoor** [comment]
@gabritto thanks for your help. Please, let me know if it's ok now.

**Review by gabritto — APPROVED** [review]

**Comment by cedoor** [comment]
Ready to merge

**Comment by stanbar** [comment]
LGTM thanks for adding types 🙏

**Comment by stanbar** [comment]
I'm missing one thing. 
Neither `Generator` nor `Base8` are exported on the `BabyJub` type.
Should I open a new PR?

**Comment by cedoor** [comment]
> Should I open a new PR?

Yes please, they should be attributes of the `BabyJub` class right?

Something like:

```ts
...

export interface BabyJub {
    Generator: Point
    Base8: Point
    
    addPoint(a: Point, b: Point): Point;

    mulPointEscalar(base: Point, e: BigNumberish): Point;

    inSubgroup(point: Point): boolean;

    inCurve(point: Point): boolean;

    packPoint(point: Point): Uint8Array;

    unpackPoint(buff: Uint8Array): Point;
}

...
```

**Pull request merged**